File: test/com/amazon/ionelement/demos/java/MutableStructFields.java
package com.amazon.ionelement.demos.java;

import com.amazon.ionelement.api.Ion;
import com.amazon.ionelement.api.StructElement;
import org.junit.jupiter.api.Test;

import java.util.Arrays;

import static com.amazon.ionelement.api.ElementLoader.loadSingleElement;
import static org.junit.jupiter.api.Assertions.assertEquals;

public class MutableStructFields {

    @Test
    void CreateUpdatedStructFromExistingStruct() {
        StructElement original = loadSingleElement("{name:\"Alice\",scores:{darts:100,billiards:15,}}").asStruct();

        StructElement expected = loadSingleElement("{name:\"Alice\",scores:{darts:100,billiards:30,pingPong:200,}}").asStruct();

        StructElement updated = original.copy(Arrays.asList(Ion.field("scores", original.get("scores").asStruct().getMutableFields()
                .add("pingPong", Ion.ionInt(200))
                .set("billiards", Ion.ionInt(30))
                .toStruct())));

        assertEquals(expected, updated);
    }
}

File: test/com/amazon/ionelement/demos/kotlin/MutableStructFields.kt
package com.amazon.ionelement.demos.kotlin

import com.amazon.ionelement.api.ionInt
import com.amazon.ionelement.api.loadSingleElement
import kotlin.test.assertEquals
import org.junit.jupiter.api.Test

class MutableStructFields {
    @Test
    fun `create updated struct from existing struct`() {
        val original = loadSingleElement(
            """
            {
                name: "Alice",               
                scores: {
                    darts: 100,
                    billiards: 15,
                }
            }
            """.trimIndent()
        ).asStruct()

        val expected = loadSingleElement(
            """
                        {
                            name: "Alice",               
                            scores: {
                                darts: 100,
                                billiards: 30,
                                pingPong: 200,
                            }
                        }
            """.trimIndent()
        ).asStruct()

        val updated = original.mutableFields.set(
            "scores",
            original["scores"].asStruct().mutableFields
                .set("pingPong", ionInt(200))
                .set("billiards", ionInt(30))
                .toStruct()
        ).toStruct()

        assertEquals(expected, updated)
    }
}
